File: agentic-loops/react_loop.py
"""
Production ReAct Loop — Python
Reason → Act → Observe, repeat until answer found.

pip install anthropic tenacity structlog
export ANTHROPIC_API_KEY=sk-ant-...
"""

import os
import re
import time
import json
import subprocess
import sys
from typing import Callable, Any
from dataclasses import dataclass, field

import structlog
from anthropic import Anthropic, APIStatusError, APIConnectionError, RateLimitError
from tenacity import retry, stop_after_attempt, wait_exponential, retry_if_exception_type

# ── Logging ───────────────────────────────────────────────────────────────────
structlog.configure(processors=[
    structlog.processors.TimeStamper(fmt="iso"),
    structlog.processors.add_log_level,
    structlog.dev.ConsoleRenderer(),
])
log = structlog.get_logger()

# ── ReAct trace types ─────────────────────────────────────────────────────────
@dataclass
class Thought:
    content: str

@dataclass
class Action:
    tool: str
    input: dict

@dataclass
class Observation:
    content: str
    is_error: bool = False

@dataclass
class Answer:
    content: str

@dataclass
class Trace:
    steps: list = field(default_factory=list)
    total_tokens: int = 0
    iterations: int = 0
    elapsed_s: float = 0.0

    def add(self, step):
        self.steps.append(step)

    def to_dict(self):
        out = []
        for s in self.steps:
            if isinstance(s, Thought):
                out.append({"type": "thought", "content": s.content})
            elif isinstance(s, Action):
                out.append({"type": "action", "tool": s.tool, "input": s.input})
            elif isinstance(s, Observation):
                out.append({"type": "observation", "content": s.content, "error": s.is_error})
            elif isinstance(s, Answer):
                out.append({"type": "answer", "content": s.content})
        return out

# ── Tool registry ─────────────────────────────────────────────────────────────
TOOL_REGISTRY: dict[str, Callable] = {}
TOOL_SCHEMAS: list[dict] = []

def tool(name: str, description: str, schema: dict):
    """Decorator to register a tool."""
    def decorator(fn: Callable):
        TOOL_REGISTRY[name] = fn
        TOOL_SCHEMAS.append({
            "name": name,
            "description": description,
            "input_schema": schema,
        })
        return fn
    return decorator

# ── Built-in tools (replace bodies with real implementations) ─────────────────

@tool(
    name="think",
    description="Use this to reason step by step before acting. Does not call any external system.",
    schema={
        "type": "object",
        "properties": {"reasoning": {"type": "string", "description": "Your chain of thought"}},
        "required": ["reasoning"],
    },
)
def think(reasoning: str) -> str:
    return "OK"

@tool(
    name="search",
    description="Search the web for current information.",
    schema={
        "type": "object",
        "properties": {
            "query": {"type": "string"},
            "num_results": {"type": "integer", "default": 5},
        },
        "required": ["query"],
    },
)
def search(query: str, num_results: int = 5) -> str:
    # Replace with: SerpAPI, Brave Search, Tavily, etc.
    raise NotImplementedError("Wire up your search API here")

@tool(
    name="fetch_url",
    description="Fetch the text content of a URL.",
    schema={
        "type": "object",
        "properties": {"url": {"type": "string"}},
        "required": ["url"],
    },
)
def fetch_url(url: str) -> str:
    import urllib.request
    with urllib.request.urlopen(url, timeout=10) as r:
        raw = r.read().decode("utf-8", errors="replace")
    # Strip HTML tags
    clean = re.sub(r"<[^>]+>", " ", raw)
    clean = re.sub(r"\s+", " ", clean).strip()
    return clean[:8000]  # truncate to avoid context overflow

@tool(
    name="run_python",
    description="Execute Python code. Returns stdout + stderr. Timeout 15s.",
    schema={
        "type": "object",
        "properties": {"code": {"type": "string"}},
        "required": ["code"],
    },
)
def run_python(code: str) -> str:
    result = subprocess.run(
        [sys.executable, "-c", code],
        capture_output=True, text=True, timeout=15
    )
    out = result.stdout or ""
    err = result.stderr or ""
    combined = (out + err).strip()
    return combined[:8000] if combined else "(no output)"

@tool(
    name="read_file",
    description="Read a file from disk.",
    schema={
        "type": "object",
        "properties": {"path": {"type": "string"}},
        "required": ["path"],
    },
)
def read_file(path: str) -> str:
    with open(path, "r", encoding="utf-8") as f:
        return f.read()

@tool(
    name="write_file",
    description="Write content to a file on disk.",
    schema={
        "type": "object",
        "properties": {
            "path": {"type": "string"},
            "content": {"type": "string"},
        },
        "required": ["path", "content"],
    },
)
def write_file(path: str, content: str) -> str:
    with open(path, "w", encoding="utf-8") as f:
        f.write(content)
    return f"Written {len(content)} bytes to {path}"

@tool(
    name="finish",
    description="Call this when you have the final answer. Ends the loop.",
    schema={
        "type": "object",
        "properties": {"answer": {"type": "string"}},
        "required": ["answer"],
    },
)
def finish(answer: str) -> str:
    return answer

# ── API call with retry ───────────────────────────────────────────────────────
client = Anthropic(api_key=os.environ["ANTHROPIC_API_KEY"], timeout=60.0, max_retries=0)

@retry(
    retry=retry_if_exception_type((APIConnectionError, RateLimitError)),
    wait=wait_exponential(multiplier=1, min=2, max=60),
    stop=stop_after_attempt(6),
)
def _call_api(messages: list, system: str) -> Any:
    return client.messages.create(
        model="claude-opus-4-6",
        max_tokens=4096,
        system=system,
        tools=TOOL_SCHEMAS,
        messages=messages,
    )

# ── ReAct system prompt ───────────────────────────────────────────────────────
REACT_SYSTEM = """You are an autonomous agent operating in a ReAct loop (Reason + Act + Observe).

For every task:
1. Use the `think` tool to reason about what to do next before acting.
2. Call the appropriate tool to act.
3. Observe the result and reason again.
4. Repeat until you have a complete, verified answer.
5. Call `finish` with your final answer when done.

Rules:
- Always think before acting. Never skip the think step.
- If a tool errors, reason about why and try a different approach.
- Do not guess. If unsure, search or fetch.
- Be thorough. Do not call finish until the task is fully complete.
"""

# ── Core ReAct engine ─────────────────────────────────────────────────────────
def run_react(
    task: str,
    system_extra: str = "",
    max_iterations: int = 30,
    on_step: Callable[[Any], None] | None = None,
) -> tuple[str, Trace]:
    """
    Run a full ReAct loop.

    Args:
        task:           The task/question to solve.
        system_extra:   Additional system instructions appended to base prompt.
        max_iterations: Hard iteration cap.
        on_step:        Optional callback called after each Thought/Action/Observation.

    Returns:
        (final_answer: str, trace: Trace)
    """
    system = REACT_SYSTEM + ("\n\n" + system_extra if system_extra else "")
    messages = [{"role": "user", "content": task}]
    trace = Trace()
    start = time.time()

    log.info("react.start", task=task[:120])

    for i in range(max_iterations):
        trace.iterations = i + 1
        log.info("react.iteration", n=i + 1)

        try:
            response = _call_api(messages, system)
        except APIStatusError as e:
            log.error("react.api_error", status=e.status_code, body=e.message)
            raise RuntimeError(f"API error {e.status_code}: {e.message}") from e

        trace.total_tokens += response.usage.input_tokens + response.usage.output_tokens
        messages.append({"role": "assistant", "content": response.content})

        if response.stop_reason == "end_turn":
            # Model responded with text instead of tool call — treat as final answer
            for block in response.content:
                if hasattr(block, "text") and block.text.strip():
                    answer = block.text.strip()
                    trace.add(Answer(answer))
                    trace.elapsed_s = round(time.time() - start, 2)
                    log.info("react.done", via="end_turn", **_trace_stats(trace))
                    return answer, trace

        if response.stop_reason == "tool_use":
            tool_results = []

            for block in response.content:
                if block.type == "tool_use":
                    tool_name = block.name
                    tool_input = block.input

                    # Record thought separately from actions
                    if tool_name == "think":
                        thought = Thought(tool_input.get("reasoning", ""))
                        trace.add(thought)
                        if on_step:
                            on_step(thought)
                        log.info("react.thought", content=thought.content[:200])
                        tool_results.append({
                            "type": "tool_result",
                            "tool_use_id": block.id,
                            "content": "OK",
                        })
                        continue

                    if tool_name == "finish":
                        answer = tool_input.get("answer", "")
                        trace.add(Answer(answer))
                        trace.elapsed_s = round(time.time() - start, 2)
                        log.info("react.done", via="finish", **_trace_stats(trace))
                        return answer, trace

                    action = Action(tool=tool_name, input=tool_input)
                    trace.add(action)
                    if on_step:
                        on_step(action)
                    log.info("react.action", tool=tool_name, input=str(tool_input)[:200])

                    # Execute tool
                    is_error = False
                    try:
                        if tool_name not in TOOL_REGISTRY:
                            raise ValueError(f"Tool '{tool_name}' not registered")
                        result = str(TOOL_REGISTRY[tool_name](**tool_input))
                    except NotImplementedError as e:
                        result = f"Tool not yet implemented: {e}. Use a different approach."
                        is_error = True
                    except Exception as e:
                        result = f"Tool error ({type(e).__name__}): {e}"
                        is_error = True
                        log.error("react.tool_error", tool=tool_name, error=str(e))

                    observation = Observation(content=result, is_error=is_error)
                    trace.add(observation)
                    if on_step:
                        on_step(observation)
                    log.info("react.observation", tool=tool_name,
                             result_len=len(result), error=is_error)

                    tool_results.append({
                        "type": "tool_result",
                        "tool_use_id": block.id,
                        "content": result,
                        "is_error": is_error,
                    })

            messages.append({"role": "user", "content": tool_results})
            continue

        log.warning("react.unexpected_stop", stop_reason=response.stop_reason)
        break

    trace.elapsed_s = round(time.time() - start, 2)
    raise RuntimeError(
        f"ReAct loop did not finish within {max_iterations} iterations. "
        f"Tokens used: {trace.total_tokens}"
    )

def _trace_stats(trace: Trace) -> dict:
    return {
        "iterations": trace.iterations,
        "tokens": trace.total_tokens,
        "elapsed_s": trace.elapsed_s,
        "steps": len(trace.steps),
    }

# ── CLI entry point ───────────────────────────────────────────────────────────
if __name__ == "__main__":
    import sys

    task = " ".join(sys.argv[1:]) if len(sys.argv) > 1 else (
        "Calculate the 47th Fibonacci number using code, then explain the result."
    )

    def print_step(step):
        if isinstance(step, Thought):
            print(f"\n💭 THOUGHT: {step.content[:300]}")
        elif isinstance(step, Action):
            print(f"\n⚡ ACTION: {step.tool}({json.dumps(step.input)[:200]})")
        elif isinstance(step, Observation):
            marker = "❌" if step.is_error else "👁"
            print(f"\n{marker} OBSERVE: {step.content[:300]}")

    answer, trace = run_react(task, on_step=print_step)

    print("\n" + "=" * 60)
    print("FINAL ANSWER:")
    print(answer)
    print("=" * 60)
    print(f"Iterations: {trace.iterations} | Tokens: {trace.total_tokens} | Time: {trace.elapsed_s}s")
    print("\nFull trace (JSON):")
    print(json.dumps(trace.to_dict(), indent=2))

File: agentic-loops/react_loop.ts
/**
 * Production ReAct Loop — TypeScript
 * Reason → Act → Observe, repeat until finish() called.
 *
 * npm install @anthropic-ai/sdk
 * export ANTHROPIC_API_KEY=sk-ant-...
 * npx ts-node react_loop.ts "your task here"
 */

import Anthropic from "@anthropic-ai/sdk";
import { execSync } from "child_process";
import * as fs from "fs";
import * as https from "https";

// ── Trace types ───────────────────────────────────────────────────────────────
type Thought    = { type: "thought";     content: string };
type Action     = { type: "action";      tool: string; input: Record<string, unknown> };
type Observation= { type: "observation"; content: string; error: boolean };
type Answer     = { type: "answer";      content: string };
type Step       = Thought | Action | Observation | Answer;

interface Trace {
  steps:       Step[];
  totalTokens: number;
  iterations:  number;
  elapsedMs:   number;
}

// ── Tool registry ─────────────────────────────────────────────────────────────
type ToolFn = (input: Record<string, unknown>) => Promise<string> | string;

interface ToolSchema {
  name:         string;
  description:  string;
  input_schema: Record<string, unknown>;
}

const toolRegistry  = new Map<string, ToolFn>();
const toolSchemas:  ToolSchema[] = [];

function registerTool(schema: ToolSchema, fn: ToolFn) {
  toolRegistry.set(schema.name, fn);
  toolSchemas.push(schema);
}

// ── Built-in tools ────────────────────────────────────────────────────────────

registerTool({
  name: "think",
  description: "Reason step by step before acting. Does not call external systems.",
  input_schema: {
    type: "object",
    properties: { reasoning: { type: "string" } },
    required: ["reasoning"],
  },
}, async ({ reasoning }) => {
  void reasoning; // logged by engine
  return "OK";
});

registerTool({
  name: "search",
  description: "Search the web for current information.",
  input_schema: {
    type: "object",
    properties: {
      query:       { type: "string" },
      num_results: { type: "integer" },
    },
    required: ["query"],
  },
}, async ({ query }) => {
  // Replace with Brave, Tavily, SerpAPI, etc.
  throw new Error(`Search not wired up. Query was: ${query}`);
});

registerTool({
  name: "fetch_url",
  description: "Fetch the text content of a URL.",
  input_schema: {
    type: "object",
    properties: { url: { type: "string" } },
    required: ["url"],
  },
}, ({ url }) => new Promise((resolve, reject) => {
  https.get(url as string, (res) => {
    let data = "";
    res.on("data", (chunk) => (data += chunk));
    res.on("end", () => {
      const clean = data.replace(/<[^>]+>/g, " ").replace(/\s+/g, " ").trim();
      resolve(clean.slice(0, 8000));
    });
  }).on("error", reject);
}));

registerTool({
  name: "run_node",
  description: "Execute JavaScript code in Node.js. Returns stdout/stderr. Timeout 15s.",
  input_schema: {
    type: "object",
    properties: { code: { type: "string" } },
    required: ["code"],
  },
}, ({ code }) => {
  try {
    const result = execSync(`node -e ${JSON.stringify(code)}`, {
      timeout: 15000,
      encoding: "utf8",
    });
    return result.trim() || "(no output)";
  } catch (e: unknown) {
    const err = e as { stdout?: string; stderr?: string; message?: string };
    return (err.stdout || "") + (err.stderr || err.message || "unknown error");
  }
});

registerTool({
  name: "read_file",
  description: "Read a file from disk.",
  input_schema: {
    type: "object",
    properties: { path: { type: "string" } },
    required: ["path"],
  },
}, ({ path }) => fs.readFileSync(path as string, "utf8"),
);

registerTool({
  name: "write_file",
  description: "Write content to a file on disk.",
  input_schema: {
    type: "object",
    properties: {
      path:    { type: "string" },
      content: { type: "string" },
    },
    required: ["path", "content"],
  },
}, ({ path, content }) => {
  fs.writeFileSync(path as string, content as string, "utf8");
  return `Written ${(content as string).length} bytes to ${path}`;
});

registerTool({
  name: "finish",
  description: "Call this when you have the final answer. Ends the loop.",
  input_schema: {
    type: "object",
    properties: { answer: { type: "string" } },
    required: ["answer"],
  },
}, ({ answer }) => answer as string,
);

// ── Client ────────────────────────────────────────────────────────────────────
const client = new Anthropic({
  apiKey: process.env.ANTHROPIC_API_KEY!,
  timeout: 60_000,
  maxRetries: 4,
});

// ── ReAct system prompt ───────────────────────────────────────────────────────
const REACT_SYSTEM = `You are an autonomous agent operating in a ReAct loop (Reason + Act + Observe).

For every task:
1. Use the think tool to reason about what to do next before acting.
2. Call the appropriate tool to act.
3. Observe the result and reason again.
4. Repeat until you have a complete, verified answer.
5. Call finish with your final answer when done.

Rules:
- Always think before acting. Never skip the think step.
- If a tool errors, reason about why and try a different approach.
- Do not guess. If unsure, search or fetch.
- Be thorough. Do not call finish until the task is fully complete.`;

// ── Core ReAct engine ─────────────────────────────────────────────────────────
export async function runReact(
  task: string,
  options: {
    systemExtra?:   string;
    maxIterations?: number;
    onStep?:        (step: Step) => void;
  } = {}
): Promise<{ answer: string; trace: Trace }> {
  const { systemExtra = "", maxIterations = 30, onStep } = options;
  const system = REACT_SYSTEM + (systemExtra ? `\n\n${systemExtra}` : "");

  const messages: Anthropic.MessageParam[] = [
    { role: "user", content: task },
  ];

  const trace: Trace = { steps: [], totalTokens: 0, iterations: 0, elapsedMs: 0 };
  const startTime = Date.now();

  console.error(`[react] START: ${task.slice(0, 120)}`);

  for (let i = 0; i < maxIterations; i++) {
    trace.iterations = i + 1;
    console.error(`[react] iteration ${i + 1}`);

    const response = await client.messages.create({
      model:      "claude-opus-4-6",
      max_tokens: 4096,
      system,
      tools:      toolSchemas as Anthropic.Tool[],
      messages,
    });

    trace.totalTokens += response.usage.input_tokens + response.usage.output_tokens;
    messages.push({ role: "assistant", content: response.content });

    if (response.stop_reason === "end_turn") {
      for (const block of response.content) {
        if (block.type === "text" && block.text.trim()) {
          const answer = block.text.trim();
          trace.steps.push({ type: "answer", content: answer });
          trace.elapsedMs = Date.now() - startTime;
          return { answer, trace };
        }
      }
    }

    if (response.stop_reason === "tool_use") {
      const toolResults: Anthropic.ToolResultBlockParam[] = [];

      for (const block of response.content) {
        if (block.type !== "tool_use") continue;

        const { id, name, input } = block;
        const inp = input as Record<string, unknown>;

        // Think step
        if (name === "think") {
          const thought: Thought = { type: "thought", content: inp.reasoning as string };
          trace.steps.push(thought);
          onStep?.(thought);
          console.error(`[react] 💭 ${thought.content.slice(0, 200)}`);
          toolResults.push({ type: "tool_result", tool_use_id: id, content: "OK" });
          continue;
        }

        // Finish
        if (name === "finish") {
          const answer = inp.answer as string;
          trace.steps.push({ type: "answer", content: answer });
          trace.elapsedMs = Date.now() - startTime;
          console.error(`[react] DONE — ${trace.iterations} iterations, ${trace.totalTokens} tokens, ${trace.elapsedMs}ms`);
          return { answer, trace };
        }

        // Regular action
        const action: Action = { type: "action", tool: name, input: inp };
        trace.steps.push(action);
        onStep?.(action);
        console.error(`[react] ⚡ ${name}(${JSON.stringify(inp).slice(0, 200)})`);

        let result: string;
        let isError = false;
        try {
          const fn = toolRegistry.get(name);
          if (!fn) throw new Error(`Tool '${name}' not registered`);
          result = await fn(inp);
        } catch (e: unknown) {
          result = `Tool error: ${(e as Error).message}`;
          isError = true;
          console.error(`[react] ❌ ${name} error: ${result}`);
        }

        const obs: Observation = { type: "observation", content: result, error: isError };
        trace.steps.push(obs);
        onStep?.(obs);
        console.error(`[react] 👁 ${result.slice(0, 200)}`);

        toolResults.push({
          type:        "tool_result",
          tool_use_id: id,
          content:     result,
          is_error:    isError,
        });
      }

      messages.push({ role: "user", content: toolResults });
      continue;
    }

    console.error(`[react] unexpected stop_reason: ${response.stop_reason}`);
    break;
  }

  trace.elapsedMs = Date.now() - startTime;
  throw new Error(
    `ReAct loop did not finish within ${maxIterations} iterations. Tokens: ${trace.totalTokens}`
  );
}

// ── CLI entry point ───────────────────────────────────────────────────────────
if (require.main === module) {
  const task = process.argv.slice(2).join(" ") ||
    "Calculate the 47th Fibonacci number using code, then explain the result.";

  runReact(task, {
    onStep(step) {
      if (step.type === "thought")     process.stdout.write(`\n💭 THOUGHT: ${step.content.slice(0, 300)}\n`);
      if (step.type === "action")      process.stdout.write(`\n⚡ ACTION: ${step.tool}(${JSON.stringify(step.input).slice(0, 200)})\n`);
      if (step.type === "observation") process.stdout.write(`\n${step.error ? "❌" : "👁"} OBSERVE: ${step.content.slice(0, 300)}\n`);
    },
  }).then(({ answer, trace }) => {
    console.log("\n" + "=".repeat(60));
    console.log("FINAL ANSWER:");
    console.log(answer);
    console.log("=".repeat(60));
    console.log(`Iterations: ${trace.iterations} | Tokens: ${trace.totalTokens} | Time: ${trace.elapsedMs}ms`);
    console.log("\nFull trace:");
    console.log(JSON.stringify(trace.steps, null, 2));
  }).catch((e) => {
    console.error("FATAL:", e.message);
    process.exit(1);
  });
}
